File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        def make_graph(node):
            if node is None:
                return
            if node.left:
                graph[node.val].append(node.left.val)
                graph[node.left.val].append(node.val)
            if node. right:
                graph[node.val].append(node.right.val)
                graph[node.right.val].append(node.val)
            make_graph(node.left)
            make_graph(node.right)
        
        graph = defaultdict(list)
        make_graph(root)
        visited=set()
        queue = deque([start])
        time = -1
        while queue:
            time+=1
            for _ in range(len(queue)):
                curr_node = queue.popleft()
                visited.add(curr_node)
                for neighbor in graph[curr_node]:
                    if neighbor not in visited:
                        queue.append(neighbor)
        return time

File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<int, vector<int>> graph;
    void makeGraph(TreeNode* root){
        if (!root) {return;}
        if (root->left){
            graph[root->val].push_back(root->left->val);
            graph[root->left->val].push_back(root->val);
        }
        if (root->right){
            graph[root->val].push_back(root->right->val);
            graph[root->right->val].push_back(root->val);
        }
        makeGraph(root->left);
        makeGraph(root->right);
    }

    
    int amountOfTime(TreeNode* root, int start) {
        makeGraph(root);

        queue<int> q;
        q.push(start);

        unordered_set<int> visited;
        visited.insert(start);

        int time = -1;
        while (!q.empty()) {
            ++time;
            int levelSize = q.size();

            for (int i = 0; i < levelSize; ++i) {
                int curr = q.front();
                q.pop();

                for (int neighbor : graph[curr]) {
                    if (!visited.count(neighbor)) {
                        q.push(neighbor);
                        visited.insert(neighbor);
                    }
                }
            }
        }
        return time;}
};